cmd/provenanced/cmd/root.go

```diff
@@ -438,18 +441,94 @@ func getIAVLCacheSize(options servertypes.AppOptions) int {
 	return iavlCacheSize
 }
 
-// isTestnetFlagSet returns true if the command was invoked with the --testnet flag.
-// This should be the same as viper.Get("testnet"), but can be used without needing
-// to set up viper.
-func isTestnetFlagSet() bool {
+// isTestnetFlagSet returns true if the args or env vars say to use testnet values.
+// It differs from viper.Get("testnet") in that this will not look into any config files.
+// That shouldn't be a problem because it's not in any of the templates. So someone would
+// need to have it in a custom.toml file, which I don't think anyone utilizes.
+//
+// We need this because the SDK requires an address codec in order to create most of their
+// cobra commands. Those commands need to be created and added to the root command well
+// before we get to the point where we're ready to read in config files. But that address
+// codec needs to have the HRP defined, which depends on the --testnet flag.
+func isTestnetFlagSet(args []string) bool {
+	// Go through the args backwards so that if multiple testnet flags are provided, the last one is used.
+	for i := len(args) - 1; i >= 0; i-- {
+		arg := args[i]
+		if arg == "-t" || arg == "--testnet" {
+			return true
+		}
+		if strings.HasPrefix(arg, "-t=") {
+			return cast.ToBool(arg[3:])
+		}
+		if strings.HasPrefix(arg, "--testnet=") {
+			return cast.ToBool(arg[10:])
+		}
+	}
 	ev := os.Getenv("PIO_TESTNET")
 	if len(ev) > 0 {
 		return cast.ToBool(ev)
 	}
-	for _, arg := range os.Args[1:] {
-		if arg == "-t" || arg == "--testnet" {
-			return true
+	return false
+}
```

Comment:
It's done manually here because it's being called in a `func init()` and a) should be as light-weigh as possible, and b) doesn't have access to all the other stuff that gets set up later (such as all the command and flag definitions). The `pflags` library isn't a good fit here because a) the flag is actually defined elsewhere, b) the ONLY flag we care about here is the `--testnet` or `-t` flag.

Also, this is a boolean flag. Boolean flags behave slightly differently from other flags in that they don't allow for the value to be provided after it as another arg. I.e. these args don't work: `-t true`, `--testnet true`. That's true of `pflags`, `cobra` and just about all of the CLI utilities in existence. So the only patterns I need to look for are `-t`, `--testnet`, `-t=<value>` and `--testnet=<value>`.

It handles all the standard boolean flag formats and it's checking all of the args provided in the invocation, so there's no chance that it misses "flags provided after subcommands." And this `--testnet` flag is persistent across all of our commands, so it is legal to have anywhere in the invocation regardless of subcommands, i.e. users can supply it at an arg position. So it is being parsed correctly regardless of its position relative to subcommands.